multi-pool: Determine IP pool based on `ipam.cilium.io/ip-pool` annotation

gandro opened the pull request:
This adds a mechanism which allows workloads to specify what IP pool their IP should be allocated from in multi-pool IPAM mode (see https://github.com/cilium/cilium/pull/22762 for details).

It will do that by first checking if the pod has a `ipam.cilium.io/ip-pool`  annotation, and if present, will return that pool to be used for IP allocation. If the pod does not have such an annotation, it will then check for the same annotation on the pod's namespace. If that is also not present, we fall back on the `default` pool.

:bulb: **Review per commit**

I have manually tested this on kind as follows:

```console
$ helm upgrade --install cilium ./install/kubernetes/cilium --namespace kube-system \
    --set debug.enabled=true \
    --set image.override=localhost:5000/cilium/cilium-dev:local \
    --set image.pullPolicy=IfNotPresent \
    --set operator.image.override=localhost:5000/cilium/operator-generic:local \
    --set operator.image.pullPolicy=IfNotPresent \
    --set operator.replicas=1 \
    --set hubble.relay.enabled=true \
    --set ipam.mode=multi-pool \
    --set "ipam.operator.multiPoolMap.default.ipv4CIDRs[0]=10.10.0.0/16" \
    --set "ipam.operator.multiPoolMap.default.ipv4MaskSize=27" \
    --set "ipam.operator.multiPoolMap.mars.ipv4CIDRs[0]=10.20.0.0/16" \
    --set "ipam.operator.multiPoolMap.mars.ipv4MaskSize=24" \
    --set "extraConfig.multi-pool-node-pre-alloc=default=64\,mars=64" \
    --set tunnel=disabled \
    --set autoDirectNodeRoutes=true \
    --set ipv4NativeRoutingCIDR=10.0.0.0/8 \
    --set endpointRoutes.enabled=true \
    --set-string extraConfig.enable-local-node-route=false \
    --set kubeProxyReplacement=strict \
    --set bpf.masquerade=true
$ k create ns cilium-test
$ k annotate ns cilium-test ipam.cilium.io/ip-pool=mars
$ cilium connectivity test
...
✅ All 41 tests (291 actions) successful, 8 tests skipped, 0 scenarios skipped.
$ k -n default get pods -o wide
NAME                         READY   STATUS    RESTARTS   AGE   IP            NODE           NOMINATED NODE   READINESS GATES
deathstar-54bb8475cc-89hwl   1/1     Running   0          9m    10.10.1.77    kind-worker2   <none>           <none>
deathstar-54bb8475cc-8gfdd   1/1     Running   0          9m    10.10.0.83    kind-worker    <none>           <none>
tiefighter                   1/1     Running   0          9m    10.10.1.103   kind-worker    <none>           <none>
xwing                        1/1     Running   0          9m    10.10.1.101   kind-worker    <none>           <none>
$ k -n cilium-test get pods -o wide
NAME                               READY   STATUS    RESTARTS   AGE     IP            NODE           NOMINATED NODE   READINESS GATES
client-7b78db77d5-dkzh6            1/1     Running   0          7m59s   10.20.0.94    kind-worker    <none>           <none>
client2-78f748dd67-j2qhg           1/1     Running   0          7m59s   10.20.0.103   kind-worker    <none>           <none>
echo-other-node-8689b5dd4f-6pk5l   2/2     Running   0          7m59s   10.20.2.182   kind-worker2   <none>           <none>
echo-same-node-5b55d8bdd7-jdb66    2/2     Running   0          7m59s   10.20.0.225   kind-worker    <none>           <none>
$ # note the different IP range for the cilium-test pods                ⬆️
```

CI and documentation will be added in a separate PR. See meta issue #25470

gandro committed:
api: Add IPAM pool name to allocated addresses

This adds the IPAM pool name to the AddressPair type which is both the
result of an IPAM allocation and an argument of the EndpointCreate
request. This allows the API client to know the pool effectively used to
allocate an IP.  The caller needs to keep track of this, as it will have
to pass the pool back to the IPAM subsystem when the IP is released.

In Cilium, the IP is typically not released not by the CNI plugin, but
directly by the endpoint package when the endpoint is deleted. This is
why the field is added to the AddressPair type (which shared between the
IPAM and endpoint types), rather than just to the IPAMAddressResponse.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
endpoint: Track IPAM pool of endpoint address

This adds a field to the endpoint structure to track where the endpoint's IP
has been allocated from. This is needed both for IP release, as well as
endpoint restoration.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
k8s: Add stream for local pod resources

This adds a new shared resource for local pods, i.e. pods scheduled on
the local node. This is useful for cells which need to access the pod
resource for a Cilium-managed endpoint.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipam: Add IPAM metadata manager

This adds a new cell which is responsible for determining the IP pool of a
particular pod. It will do that by first checking if the pod has the `ip-pool`
annotation, and if present, will return that pool to be used for IP allocation.
If the pod does not have such an annotation, it will then check for the same
annotation on the namespace. If that is also not present, we fall back on the
`default` pool.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
annotation: Add `ipam.cilium.io/ip-pool`

This adds an annotation which may be added to pods or namespaces to determine
the IP pool should be used for those workloads.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

oblazek commented on pkg/annotation/k8s.go:
```diff
@@ -24,6 +24,9 @@ const (
 	// ServicePrefix is the common prefix for service related annotations.
 	ServicePrefix = "service.cilium.io"
 
+	// ServicePrefix is the common prefix for IPAM related annotations.
```
```suggestion
	// IPAMPrefix is the common prefix for IPAM related annotations.
```

gandro replied:
Thanks! Fixed

gandro committed:
ipam: Hook up IPAM metadata manager

This commit introduces the IPAM metadata manager to the IPAM subsystem.
When allocating a new random IP via `AllocateNext` with
an empty pool name, the IPAM subsystem will now call into the IPAM
metadata manager to obtain the pool name for the IP owner. If a pool is
provided by the caller, then no lookup is performed and the caller's
pool takes precedence.

For `ReleaseIP` and `AllocateIP`, we now also expect the IP pool to be
explicitly provided. `AllocateIP` is only used for IP restoration, in
which case we expect the caller to know what pool to restore the IP from
(because restoring the IP from a different pool than it was originally
allocated from might lead to bugs). For `ReleaseIP`, we don't actually
know the owner name, so we also expect it to be provided, because again,
releasing an IP into a different pool than then one it was allocated
from might lead to bugs.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipam: Remove unused ability to release IPs by owner name

This removes a functionality which was added for a bug fix in
03df4a519c503d090f6dfc428b9da31b42eba251, but then made obsolete in
82a8c71556820af2fd74329ce68169c25f4ce00b when the bug fix turned out to
be insufficient.

There are no other users of this funcionality. This commit also moves
the parsing of the IP address back into the API handler. This ensures
that the API actually returns the correct (i.e. documented) HTTP error
code when the IP is invalid.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

christarazi commented on pkg/ipam/metadata/manager.go:
```diff
@@ -0,0 +1,137 @@
+	if r.Namespace != "" {
+		name = r.Namespace + "/" + r.Name
+	}
+	return fmt.Sprintf("Resource %s %q not found", r.Resource, name)
```
Nit: error strings should not be capitalized

gandro replied:
Thanks! Good catch! Will address once more reviews are in

christarazi reviewed (approved):
Well-done commits, very clean history! Makes it a breeze to review. Nothing major from my side.

doniacld commented on pkg/ipam/allocator.go:
```diff
@@ -56,6 +69,13 @@ func (ipam *IPAM) allocateIP(ip net.IP, owner string, pool Pool, needSyncUpstrea
 	ipam.allocatorMutex.Lock()
 	defer ipam.allocatorMutex.Unlock()
 
+	if pool == "" {
```
You could add a comment to be explicit that it should not happen.

gandro replied:
Yeah, so for context for other reviewers: `allocateIP` is usually only called in the context of endpoint restoration, in which point we assume `pool` to be non-empty, so this is basically dead code.

So maybe we should error out here if someone calls `allocateIP` without an explicit pool. Endpoint restoration should call it with an explicit pool (even if it's just `default`). And if the function is called via Cilium API (which I don't think anyone does), there is actually no way to inspect the pool which was selected.

So yeah, discussing this with Donia I'm coming to the conclusion that we probably should just error out if no explicit pool was provided. We can always relax that restriction later.

gandro replied:
Added an error instead and updated the commit message the reflect that.

nathanjsweet reviewed (approved):
LGTM for API

doniacld commented on pkg/ipam/metadata/manager.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+type Manager struct {
+	namespaceResource resource.Resource[*slim_core_v1.Namespace]
```
Already asked verbally but might be worth putting in public :)
Since `namespaceResource` is not really used, do we need to store it?

gandro replied:
Yes, we need `namespaceResource` (and it's `Pod` counterpart) to get access to the`namespaceStore`, which is what we actually want. Unfortunately, because obtaining the `Store` is a blocking operation (ref), we cannot obtain the store in the store in the `newIPAMMetadataManager` constructor and thus need to store the reference in the struct until `OnStart()` is called, at which point we are then allowed to obtain the `Store` reference.

My understanding of the hive framework is the daemon consumer of `ipamMetadata.Manager` should not observe a reference to it until it has been properly "started", so delaying the initialization of the stores into the `Start` hook is safe.

doniacld commented on daemon/cmd/state.go:
```diff
@@ -363,20 +363,22 @@ func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) error {
 	var err error
```
For the history, it is more robust for a defer statement to define the variable as the returned value to make even though other operations are done in the middle, the defer is actually executed properly.

```suggestion
func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) (err error) {
    if option.Config.EnableIPv6 && ep.IPv6.IsValid() {
```

gandro replied:
Fixed in https://github.com/cilium/cilium/pull/25511/commits/6bdaabaffcbf662c87f177ae3894e83725e761e8

tklauser commented on api/v1/openapi.yaml:
```diff
@@ -1686,12 +1680,18 @@ definitions:
         description: UUID of IPv4 expiration timer
         type: string
+      ipv4-pool-name:
+        description: IPAM pool from which this IP has been allocated from
```
nit:
```suggestion
        description: IPAM pool from which this IPv4 address was allocated
```

tklauser commented on api/v1/openapi.yaml:
```diff
@@ -1686,12 +1680,18 @@ definitions:
         description: UUID of IPv6 expiration timer
         type: string
+      ipv6-pool-name:
+        description: IPAM pool from which this IP has been allocated from
```
nit:

```suggestion
        description: IPAM pool from which this IPv6 was allocated
```

tklauser commented on pkg/endpoint/endpoint.go:
```diff
@@ -166,9 +166,15 @@ type Endpoint struct {
 	// IPv4 is the IPv4 address of the endpoint
 	IPv4 netip.Addr
 
+	// IPv4IPAMPool is the IPAM address pool from which the IPv4 address has been allocated from
```
```suggestion
	// IPv4IPAMPool is the IPAM address pool from which the IPv4 address was allocated
```

tklauser commented on pkg/endpoint/endpoint.go:
```diff
@@ -166,9 +166,15 @@ type Endpoint struct {
 	// IPv6 is the IPv6 address of the endpoint
 	IPv6 netip.Addr
 
+	// IPv6IPAMPool is the IPAM address pool from which the IPv6 address has been allocated from
```
```suggestion
	// IPv6IPAMPool is the IPAM address pool from which the IPv6 address was allocated
```

tklauser reviewed (approved):
Nice work! Thanks for splitting it up into self-contained commits which made it easy to review. Some comment/doc wording nits only, feel free to ignore.

gandro committed:
ipam: Make allocateIPsLocked a bit more robust

This commit contains no functional changes, but moves the variable
definition of `err` from a local variable to a named return value. This
makes a difference when reading `err` in the `defer` block, because if
we ever return with an error without also setting `err` to a non-nil
value, the `defer` statement which releases the IPv6 address would not
release it.

At the moment, the implementation never returns an error without also
setting `err` explicitly, but moving the variable definition into the
function signature ensures that future code changes may not introduce
subtle bugs.

See the following example for details: https://go.dev/play/p/3l7FDg5k6yP

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

lmb commented on daemon/cmd/ipam.go:
```diff
@@ -554,6 +561,9 @@ func (d *Daemon) startIPAM() {
 	log.Info("Initializing node addressing")
 	// Set up ipam conf after init() because we might be running d.conf.KVStoreIPv4Registration
 	d.ipam = ipam.NewIPAM(d.datapath.LocalNodeAddressing(), option.Config, d.nodeDiscovery, d.k8sWatcher, &d.mtuConfig, d.clientset)
+	if d.ipamMetadata != nil {
```
If I read the code correctly, d.ipamMetadata is already created via hive? Why add a fallback here, and not where the ipamMetaddata is initialized?

gandro replied:
The hive `ipamMetaddata` constructor will return nil if it's running in an unsupported IPAM mode:
https://github.com/cilium/cilium/blob/0d4ba6a6d4edb21c92f6e469566d4d408c2ac31b/pkg/ipam/metadata/cell.go#L34-L36

And I do not want to pass in a potentially `nil` value as an interface type into `WithMetadata` because it will break the `nil` checks here (i.e. they will not detect that metadata is nil due to [nil interface types](https://trstringer.com/go-nil-interface-and-interface-with-nil-concrete-value/)):
https://github.com/cilium/cilium/blob/0d4ba6a6d4edb21c92f6e469566d4d408c2ac31b/pkg/ipam/allocator.go#L34-L36

Example: https://go.dev/play/p/z8C_cU0eMdU

lmb reviewed (approved):
Yup! I was out for a long weekend, sorry.
